include/vkwrap/vkwrap/mman.h

```diff
@@ -0,0 +1,556 @@
+    } else if ( old_layout == ImageLayout::eUndefined && new_layout == ImageLayout::eDepthStencilAttachmentOptimal )
+    {
+        src = PipelineStageFlagBits::eTopOfPipe;
+        dst = PipelineStageFlagBits::eEarlyFragmentTests;
+    } else
+    {
+        throw Error{ "choosePipelineStages: unimplemented layout transition." };
+    }
+
+    return { src, dst };
+} // choosePipelineStages
+
+class Mman
+{
+
+  private:
+    struct BufferInfo
+    {
+        VmaAllocation allocation;
+        vk::DeviceSize size;
+    }; // struct BufferInfo
+
+    struct ImageInfo
+    {
+        VmaAllocation allocation;
+        vk::Format format;
+        vk::ImageLayout layout;
+        vk::Extent3D extent;
+    }; // struct ImageInfo
+
+    class OneTimeCommand : private vk::UniqueCommandBuffer
+    {
+
+      private:
+        using Base = vk::UniqueCommandBuffer;
+
+      private:
+        vk::Queue m_queue;
+
+        vk::UniqueCommandBuffer createCommandBuffer( vk::Device device, vk::CommandPool cmd_pool ) const
+        {
+            vk::CommandBufferAllocateInfo alloc_info{
+                .commandPool = cmd_pool,
+                .level = vk::CommandBufferLevel::ePrimary,
+                .commandBufferCount = 1 };
+
+            return std::move( device.allocateCommandBuffersUnique( alloc_info )[ 0 ] );
+        } // createCommandBuffer
+
+        void begin() const
+        {
+            // clang-format off
+            vk::CommandBufferBeginInfo begin_info{ 
+                .flags = vk::CommandBufferUsageFlagBits::eOneTimeSubmit };
+            // clang-format on
+
+            Base::get().begin( begin_info );
+        } // begin
+
+        void end() const { Base::get().end(); }
+
+      public:
+        OneTimeCommand( vk::Device device, vk::CommandPool cmd_pool, vk::Queue queue )
+            : Base{ createCommandBuffer( device, cmd_pool ) },
+              m_queue{ queue }
+        {
+            begin();
+        } // oneTimeCommand
+
+        void submitAndWait() const
+        {
+            end();
+
+            // clang-format off
+            vk::SubmitInfo submit_info{ 
+                .commandBufferCount = 1, 
+                .pCommandBuffers = &Base::get() };
+            // clang-format on
+
+            m_queue.submit( std::array{ submit_info } );
+            m_queue.waitIdle();
+        } // submitAndWait
+
+        using Base::operator->;
+    }; // class OneTimeCommand
+
+  private:
+    VmaAllocator m_vma;
+    vk::Device m_device;
+    vk::Queue m_queue;
+    vk::CommandPool m_cmd_pool;
+
+    std::unordered_map<vk::Buffer, BufferInfo> m_buffers_info;
+    std::unordered_map<vk::Image, ImageInfo> m_images_info;
+
+    BufferInfo* findInfo( vk::Buffer buffer )
+    {
+        if ( auto entry = m_buffers_info.find( buffer ); entry == m_buffers_info.end() )
+        {
+            throw Error{ "Mman: buffer info not found." };
+        } else
+        {
+            return &entry->second;
+        }
+    } // findInfo
```

Comment:
Why does this function return a pointer? Can it be null? Then there is no need to throw an exception: just return a ```nullptr```. However, it would be better in my opinion to return by value.

Reply:
If `vk::Buffer` haven't been created with `Mman::create` then there is no `BufferInfo` for it. If such situation is happened, then user is using `Mman` incorrectly. We could put here `assert(...); std::terminate();`, but in release mode `std::terminate()` won't tell any additional information.

Reply:
Talking to 'return by value'. Updating information in `BufferInfo` using pointer semantic is more clear, than accessing fields by reference.

Reply:
Alright, but still, can ```findInfo``` return a nullptr? Seems like a no to me. Then this should be a function returning a reference (an ref-annotated one at that)